src/github.com/getlantern/context/context.go

```diff
@@ -0,0 +1,232 @@
+// if necessary.
+func Enter() *Context {
+	id := curGoroutineID()
+	allmx.Lock()
+	c := contexts[id]
+	if c == nil {
+		c = makeContext(id, nil, nil)
+		contexts[id] = c
+		allmx.Unlock()
+		return c
+	}
+	allmx.Unlock()
+	return c.Enter()
+}
+
+// Enter enters a new level on this Context stack.
+func (c *Context) Enter() *Context {
+	c.mx.RLock()
+	id := c.id
+	c.mx.RUnlock()
+	next := makeContext(id, c, nil)
+	allmx.Lock()
+	contexts[id] = next
+	allmx.Unlock()
+	return next
+}
+
+// Go starts the given function on a new goroutine.
+func (c *Context) Go(fn func()) {
+	go func() {
+		id := curGoroutineID()
+		next := makeContext(id, nil, c)
+		allmx.Lock()
+		contexts[id] = next
+		allmx.Unlock()
+		fn()
+		// Clean up the context
+		allmx.Lock()
+		delete(contexts, id)
+		allmx.Unlock()
+	}()
+}
+
+// Go starts the given function on a new goroutine but sharing the context of
+// the current goroutine (if it has one).
+func Go(fn func()) {
+	c := currentContext()
+	if c != nil {
+		c.Go(fn)
+	} else {
+		go fn()
+	}
+}
+
+func makeContext(id uint64, parent *Context, branchedFrom *Context) *Context {
+	return &Context{
+		id:           id,
+		parent:       parent,
+		branchedFrom: branchedFrom,
+		data:         make(Map),
+	}
+}
+
+// Exit exits the current level on this Context stack.
+func (c *Context) Exit() *Context {
+	c.mx.RLock()
+	id := c.id
+	parent := c.parent
+	c.mx.RUnlock()
+	if parent == nil {
+		allmx.Lock()
+		delete(contexts, id)
+		allmx.Unlock()
+		return nil
+	}
+	allmx.Lock()
+	contexts[id] = parent
+	allmx.Unlock()
+	return parent
+}
+
+// Put puts a key->value pair into the current level of the context stack.
+func (c *Context) Put(key string, value interface{}) *Context {
+	c.mx.Lock()
+	c.data[key] = value
+	c.mx.Unlock()
+	return c
+}
+
+// PutGlobal puts the given key->value pair into the globalc context.
+func PutGlobal(key string, value interface{}) {
+	allmx.Lock()
+	global[key] = value
+	allmx.Unlock()
+}
+
+// PutDynamic puts a key->value pair into the current level of the context stack
+// where the value is generated by a function that gets evaluated at every Read.
+func (c *Context) PutDynamic(key string, valueFN func() interface{}) *Context {
+	value := &dynval{valueFN}
+	c.mx.Lock()
+	c.data[key] = value
+	c.mx.Unlock()
+	return c
+}
+
+// PutGlobalDynamic puts a key->value pair into the global context wwhere the value is
+// generated by a function that gets evaluated at every Read.
+func PutGlobalDynamic(key string, valueFN func() interface{}) {
+	value := &dynval{valueFN}
+	allmx.Lock()
+	global[key] = value
+	allmx.Unlock()
+}
+
+// AsMap returns a map containing all values along the stack.
+func (c *Context) AsMap() Map {
+	m := make(Map)
+	c.fill(m)
+	return m
+}
+
+func (c *Context) fill(m Map) {
+	for ctx := c; ctx != nil; {
+		ctx.mx.RLock()
+		fill(m, ctx.data)
+		next := ctx.parent
+		if next == nil {
+			next = ctx.branchedFrom
```

Comment:
We can't merge them.  We use parent when unwinding the stack on the current goroutine.  Once we reach parent == nil, we stop unwinding the stack.  Each goroutine needs to have its own stack, because parent goroutines can terminate before the child ones.